### ISSUE TITLE
Compatibility between Dedukti 2.7 and Lambdapi

### ISSUE THREAD

**Issue by francoisthire** [Deducteam/lambdapi#315]
I raise one more time this issue but I think it would be really nice to actually do something about it. 

So I recall that today, there are two main implementations of Dedukti which are:
- Dedukti (aka Dedukti 2.7) : https://github.com/Deducteam/Dedukti
- Lambdapi (aka Dedukti 3.0) : https://github.com/Deducteam/lambdapi

The problem is: If the development of Dedukti 2.7 is dropped, what will happen of tools using Dedukti 2.7?

Fortunately, for many tools, it is only a matter of syntax and except for one feature (brackets), it seems that the syntax of Dedukti 2.7 is a subset of the syntax of Dedukti 3.0 (old or new). 

The real problem is for tools that uses Dedukti 2.7 as an OCaml library. IIRC, they are five:
- Dkmeta
- Logipedia (the backend part that produces JSON files and proof files that are exported)
- SCT
- Sukerujo (used by Focalize)
- Universo

From what I know, a way to get around this issue on the short term is to have a full compatibility of the syntax between Dedukti 2.7 and Dedukti 3.0 . To achieve this part, I think the easiest way is to have a printer of Dedukti 3.0 to Dedukti 2.7 . This printer could be itself a tool around Dedukti 3.0 maybe? 

On the long term though, this solution may be not well suited because the development of Dedukti 2.7 will be dropped. On the long term, we can rewrite the code from scratch of these tools for Lambdapi, but given the experience of the past two years, I suspect this will not happen any time soon. 

Some effort has been made to have a proper API for Dedukti 2.7 . Maybe an intermediate step before translating the code of these tools from Dedukti 2.7 to Dedukti 3.0 would be to see whether the API we have proposed for Dedukti 2.7 can be implemented (again, as a tool) for Dedukti 3.0 ?

To do so, the three main interfaces that need to be implemented are:

- The top-level commands: https://github.com/Deducteam/Dedukti/blob/another-api/parsing/entry.mli
- The environment (and interface over the kernel of Dedukti 2.7) : https://github.com/Deducteam/Dedukti/blob/another-api/api/env.mli
- The processors (a generic way to process top-level commands using the environment) : https://github.com/Deducteam/Dedukti/blob/another-api/api/processor.mli

This API is not finished but almost. Only the first two interfaces are really important, I think that the last one can be in 90% built on the other two.

So to sum up, I have the impression that to increase the compatibility between the two tools, there are two solutions:
- Having a printer of Dedukti 3.0 towards the old syntax. And if the development team of Dedukti 3.0 does not want to maintain this code, to see whether this could be implemented on top of Dedukti 3.0 using Dedukti 3.0 as an OCaml library
- Having a common API between Dedukti 3.0 and Dedukti 2.7 which again could be implemented using Dedukti 3.0 as an OCaml library

What are your opinions about this? Of course, I am willing to contribute to implement these things if it is possible (on my spare time).

**Comment by Rehan-MALAK** [comment]
I know here, we are more talking about API than syntax and parser, but just for the record, here is what lambdapi legacy parser gives on dedukti "OK" tests.

```
$ find ./tests/OK -type f -name "*.dk" -exec lambdapi check {} \; 2>&1 | grep 31m | awk -F'OK' '{print $2}'
/inferingKindForType.dk, 4:0-1] Unexpected characters [#].
/guard1.dk, 9:15-16] Unexpected characters [{].
/nsteps3.dk, 19:12] Unexpected token [,].
/SR_sat_eq2.dk, 20:0-53] Cannot solve f (f ?Y) ≡ f (g ?X)
/brackets4.dk, 13:6-7] Unexpected characters [{].
/rule_name2.dk, 9:0-1] Unexpected characters [{].
/nsteps2.dk, 15:5-12] Invalid command configuration.
/special_idents1.dk, 3:0-1] Unexpected characters [{].
/guard3.dk, 9:15-16] Unexpected characters [{].
/brackets1.dk, 8:8-9] Unexpected characters [{].
/dep_B.dk, 3:4-11] No module [dep_A] required.
/rule_name.dk, 9:0-1] Unexpected characters [{].
/check/conv_command.dk, 8:0-1] Unexpected characters [#].
/check/type_command.dk, 7:0-1] Unexpected characters [#].
/brackets0b.dk, 12:5-6] Unexpected characters [{].
/brackets3.dk, 8:12-13] Unexpected characters [{].
/nsteps1.dk, 6:5-12] Invalid command configuration.
/nsteps4.dk, 10:5-16] Invalid command configuration.
/fixpoints.dk, 151:39-40] Unexpected characters [{].
/inferingKindForArrowWithCodomainType.dk, 5:0-1] Unexpected characters [#].
/brackets5.dk, 11:12-13] Unexpected characters [{].
/ho_nonlinearity.dk, 12:0-23:2] Assertion failed.
/brackets0.dk, 10:5-6] Unexpected characters [{].
/arities.dk, 19:0-25] Assertion failed.
/ho_bug2.dk, 13:0-1] Unexpected characters [#].
/special_idents2.dk, 3:20-21] Unexpected characters [{].
/brackets2.dk, 8:15-16] Unexpected characters [{].
/injective_smb.dk, 5:10] Unexpected token [eta].
/guard2.dk, 9:15-16] Unexpected characters [{].
/let_syntax.dk, 14:9] Unexpected token [:=].
```

**Comment by fblanqui** [comment]
Solution 1 has been implemented in #801. Moreover, the dk parser of LP is now in sync with the parser of dkcheck thanks to #755.
